File: conformance/runner/typescript/delay-gaps.ts
/**
 * The `delayBetweenRequests` assertion contract, kept apart from the runner so
 * its bounds branches are unit-testable (delay-gaps.test.ts).
 */

/**
 * Validates one assertion against the recorded request times, returning
 * `undefined` when it holds and a failure message otherwise.
 *
 * Gap i is the interval between request i and request i+1, so N requests yield
 * N-1 gaps. The contract in conformance/schema.json:
 *
 * - A NAMED index selects exactly that gap, bounds-checked unconditionally. A
 *   gap the run never produced is a failure, not a silent pass — the whole
 *   point of a timing pin is to catch a dropped backoff, and a dropped backoff
 *   is precisely what removes the gap.
 * - An OMITTED index requires the minimum on EVERY gap. Zero gaps means
 *   nothing was measured, so that fails too: an "every gap" rule with no gaps
 *   left would otherwise wave through a run that dropped every retry.
 * - Negative indexes are rejected rather than wrapping to the end the way the
 *   per-request assertions do. There is no sensible "last gap" when the point
 *   of naming one is to pin a specific backoff.
 *
 * `slackMs` exists because Node's timers may fire marginally BEFORE the
 * requested delay — libuv rounds the deadline down internally, so a 2000ms
 * sleep can legitimately elapse in 1999.87ms. That is a runtime property, not
 * an SDK behaviour: the SDK asked for the full interval. The Go runner never
 * sees it because Go's timers do not fire early. A sub-millisecond allowance
 * cannot mask a real regression, which would miss by hundreds of milliseconds
 * (a dropped Retry-After means ~1000ms of backoff instead of 2000ms) or by the
 * whole interval (no delay at all).
 */
export function checkDelayGaps(
  requestTimes: number[],
  minDelayMs: number | undefined,
  index: number | undefined,
  slackMs = 0,
): string | undefined {
  // An absent or zero minimum still asserts that the gap EXISTS. The default
  // lands HERE rather than at the call site so a truthiness gate cannot
  // quietly reduce the assertion to nothing.
  const min = minDelayMs ?? 0;
  const gaps = requestTimes.length - 1;
  const floor = min - slackMs;

  const short = (gap: number): string | undefined => {
    const delay = requestTimes[gap + 1]! - requestTimes[gap]!;
    return delay < floor
      ? `expected delay >= ${min}ms at gap ${gap} (allowing ${slackMs}ms timer slack), got ${delay}ms`
      : undefined;
  };

  if (index !== undefined) {
    if (index < 0) {
      return `delayBetweenRequests gap index must be non-negative, got ${index}`;
    }
    if (index >= gaps) {
      return `expected a delay at gap ${index}, but only ${requestTimes.length} request(s) were made`;
    }
    return short(index);
  }

  if (gaps < 1) {
    return `expected a delay between requests, but only ${requestTimes.length} request(s) were made`;
  }
  for (let i = 0; i < gaps; i++) {
    const failure = short(i);
    if (failure !== undefined) return failure;
  }
  return undefined;
}

File: conformance/runner/go/delay_gaps.go
package main

import (
	"fmt"
	"time"
)

// checkDelayGaps validates one `delayBetweenRequests` assertion against the
// recorded request times, returning "" when it holds and a failure message
// otherwise.
//
// Gap i is the interval between request i and request i+1, so N requests
// yield N-1 gaps. The contract in conformance/schema.json:
//
//   - A NAMED index selects exactly that gap. It is bounds-checked
//     unconditionally: a gap the run never produced is a failure, not a
//     silent pass. The whole point of a timing pin is to catch a dropped
//     backoff, and a dropped backoff is precisely what removes the gap.
//   - An OMITTED index requires the minimum on EVERY gap. Zero gaps means
//     nothing was measured, so an omitted-index assertion on a single-request
//     run fails too — the same reasoning, one step further: if every retry
//     were dropped, an "every gap" rule with no gaps left would otherwise
//     wave the run through.
//   - Negative indexes are rejected rather than wrapping to the end the way
//     headerPresent and friends do. There is no sensible "last gap" when the
//     point of naming one is to pin a specific backoff.
//
// The bounds test compares against the last valid gap rather than adding one
// to the index: `index+1 >= len(requestTimes)` overflows for math.MaxInt and
// wraps negative, sailing through the guard into an out-of-range read.
func checkDelayGaps(requestTimes []time.Time, minDelay time.Duration, index *int) string {
	gaps := len(requestTimes) - 1

	if index != nil {
		gap := *index
		if gap < 0 {
			return fmt.Sprintf("delayBetweenRequests gap index must be non-negative, got %d", gap)
		}
		if gap >= gaps {
			return fmt.Sprintf("Expected a delay at gap %d, but only %d request(s) were made", gap, len(requestTimes))
		}
		if delay := requestTimes[gap+1].Sub(requestTimes[gap]); delay < minDelay {
			return fmt.Sprintf("Expected delay >= %v at gap %d, got %v", minDelay, gap, delay)
		}
		return ""
	}

	if gaps < 1 {
		return fmt.Sprintf("Expected a delay between requests, but only %d request(s) were made", len(requestTimes))
	}
	for i := 0; i < gaps; i++ {
		if delay := requestTimes[i+1].Sub(requestTimes[i]); delay < minDelay {
			return fmt.Sprintf("Expected delay >= %v at gap %d, got %v", minDelay, i, delay)
		}
	}
	return ""
}
